python/taichi/lang/transformer.py

```diff
@@ -16,26 +17,89 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         local = self.scopes[-1]
         if self.stmt_block is not None:
             for var in reversed(local):
-                stmt = ASTTransformer.parse_stmt('del var')
+                stmt = ASTTransformerBase.parse_stmt('del var')
                 stmt.targets[0].id = var
                 self.stmt_block.append(stmt)
         self.scopes.pop()
 
 
-# Single-pass transform
-class ASTTransformer(ast.NodeTransformer):
+# Total transform
+# TODO: ASTTransformerBase -> ASTTransformer
+# TODO: ASTTransformer -> ASTTransformerTotal
+class ASTTransformer(object):
+    def __init__(self, func=None, *args, **kwargs):
+        self.pass_LowerAST = ASTTransformer_LowerAST(func=func,
+                                                     *args,
+                                                     **kwargs)
+        self.pass_Checks = ASTTransformer_Checks(func=func)
+
+    @staticmethod
+    def print_ast(tree, title=None):
+        if not impl.get_runtime().print_preprocessed:
+            return
+        if title is not None:
+            print(f'{title}:')
+        import astor
+        print(astor.to_source(tree.body[0], indent_with='  '))
+
+    def visit(self, tree):
+        self.print_ast(tree, 'Initial AST')
+        self.pass_LowerAST.visit(tree)
+        self.print_ast(tree, 'Lower AST')
+        self.pass_Checks.visit(tree)
+        self.print_ast(tree, 'Checks')
+        ast.fix_missing_locations(tree)
+        self.print_ast(tree, 'Final AST')
+
+
+class ASTTransformerBase(ast.NodeTransformer):
+    def __init__(self, func):
+        super().__init__()
+        self.func = func
+
+    @staticmethod
+    def parse_stmt(stmt):
+        return ast.parse(stmt).body[0]
+
+    @staticmethod
+    def parse_expr(expr):
+        return ast.parse(expr).body[0].value
+
+    @staticmethod
+    def func_call(name, *args):
+        return ast.Call(func=ASTTransformerBase.parse_expr(name).value,
+                        args=list(args),
+                        keywords=[])
+
+    @staticmethod
+    def make_constant(value):
+        # Do not use ast.Constant which does not exist in python3.5
+        node = ASTTransformerBase.parse_expr('0')
+        node.value = value
+        return node
+
+    @staticmethod
+    def get_targets(node):
+        if isinstance(node.target, ast.Name):
+            return [node.target.id]
+        else:
+            assert isinstance(node.target, ast.Tuple)
+            return [name.id for name in node.target.elts]
+
+
+# First-pass transform
+class ASTTransformer_LowerAST(ASTTransformerBase):
```

Comment:
Maybe rename this to `ASTTransformerPreprocess`, given we already have a `lower_ast` pass in C++?